src/System.Text.Json/src/System/Text/Json/JsonDocument.TryGetProperty.cs

```diff
@@ -0,0 +1,185 @@
+                        utf8Name = tmpUtf8.AsSpan(0, len);
+
+                        return TryGetNamedPropertyValue(
+                            candidateIndex,
+                            endIndex,
+                            utf8Name,
+                            out value);
+                    }
+                    finally
+                    {
+                        // While property names aren't usually a secret, they also usually
+                        // aren't long enough to end up in the rented buffer transcode path.
+                        //
+                        // On the basis that this is user data, go ahead and clear it.
+                        utf8Name.Clear();
+                        ArrayPool<byte>.Shared.Return(tmpUtf8);
+                    }
+                }
+
+                // Move to the previous value
+                candidateIndex -= DbRow.Size;
+            }
+
+            // None of the property names were within the range that the UTF-8 encoding would have been.
+            value = default;
+            return false;
+        }
+
+        internal bool TryGetNamedPropertyValue(int index, ReadOnlySpan<byte> propertyName, out JsonElement value)
+        {
+            CheckNotDisposed();
+
+            DbRow row = _parsedData.Get(index);
+
+            CheckExpectedType(JsonTokenType.StartObject, row.TokenType);
+
+            // Only one row means it was EndObject.
+            if (row.NumberOfRows == 1)
+            {
+                value = default;
+                return false;
+            }
+
+            int endIndex = checked(row.NumberOfRows * DbRow.Size + index);
+
+            return TryGetNamedPropertyValue(
+                index + DbRow.Size,
+                endIndex,
+                propertyName,
+                out value);
+        }
+
+        private bool TryGetNamedPropertyValue(
+            int startIndex,
+            int endIndex,
+            ReadOnlySpan<byte> propertyName,
+            out JsonElement value)
+        {
+            DbRow row;
+            ReadOnlySpan<byte> documentSpan = _utf8Json.Span;
+
+            // Move to the row before the EndObject
+            int index = endIndex - DbRow.Size;
+
+            while (index > startIndex)
+            {
+                row = _parsedData.Get(index);
+                Debug.Assert(row.TokenType != JsonTokenType.PropertyName);
+
+                // Move before the value
+                if (row.IsSimpleValue)
+                {
+                    index -= DbRow.Size;
+                }
+                else
+                {
+                    Debug.Assert(row.NumberOfRows > 0);
+                    index -= DbRow.Size * (row.NumberOfRows + 1);
+                }
+
+                row = _parsedData.Get(index);
+                Debug.Assert(row.TokenType == JsonTokenType.PropertyName);
+
+                ReadOnlySpan<byte> currentPropertyName = documentSpan.Slice(row.Location, row.SizeOrLength);
```

Comment:
Does this need to be unescaped first before doing a byte-by-byte comparison with the input property name?

Reply:
Yep, good point.